src/resources/Catalogs/ProductListingInterfaces.ts

```diff
@@ -9,11 +8,11 @@ export interface FieldValueModel {
     /**
      * The field type.
      */
-    type: string;
+    type: FieldValueType;
     /**
-     * The field value enum type.
+     * The value of the field.
      */
-    value: FieldValueType;
+    value: string | number | string[] | number[];
```

Comment:
Looks to me like you could do one better and infer the value type from the type using generics.
Example:
```ts
enum FieldValueType {
    RANGE = 'RANGE',
    STRING = 'STRING',
    DECIMAL = 'DECIMAL',
    ARRAY = 'ARRAY',
    HIERARCHIC_MULTI_VALUE = 'HIERARCHIC_MULTI_VALUE',
}

interface ValueTypeForFieldValueType {
    [FieldValueType.RANGE]: number[],
    [FieldValueType.STRING]: string,
    [FieldValueType.DECIMAL]: number,
    [FieldValueType.ARRAY]: string[],
    [FieldValueType.HIERARCHIC_MULTI_VALUE]: string[],
}


interface FieldValueModelGeneric<T extends FieldValueType> {
    type: T,
    value: ValueTypeForFieldValueType[T]
}

type FieldValueModel =
    |   FieldValueModelGeneric<FieldValueType.RANGE>
    |   FieldValueModelGeneric<FieldValueType.STRING>
    |   FieldValueModelGeneric<FieldValueType.DECIMAL>
    |   FieldValueModelGeneric<FieldValueType.ARRAY>
    |   FieldValueModelGeneric<FieldValueType.HIERARCHIC_MULTI_VALUE>

// See invalid
const invalid: FieldValueModel = {
    type: FieldValueType.RANGE,
    value: ['foo']
}

// See valid
const valid: FieldValueModel = {
    type: FieldValueType.RANGE,
    value: [1]
}
```
[See example on TS Playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAYglsANgEwGoENFmAFQJ4AOwUA3gFBSVQBKAggHIDiAolALxQDkdTznANBSoBlHNQCSTdl1ESmAoZQAizAMLiAsrQAy0zivVbtCqlFrU6ATT3mrJqgAlxzOtVVPVAfQ0BVbTnFPVB0fVg5OJxdzd3EvX39A4O1QhQBfMjI4EAAXYAAnADN0AGNiDCxcQmAYAHs8+CQ0TGx8IlJFKABtBpRylqqAOh4WAF0ALihwCAAjfM6RwVNuhF7myqIB2UlGcagAZ2y8rIBzRaplxr714AGDTR1dqdm8s8oL1YrWm9taS12Do4gY7zV5dHpNT6DSKuGJxPwBIIhZj-Q4nEFkdIZLK5QolYjgq4aGrIJCMUD5ODFAA8OCgwAAHrkQMg9rAVhD+kQAHztUzZKoTHCggBuawmVy+tXq7IlVU6OBGGIy-LaBLWRJJiHYHQAPlQ1RUNaTyUdqQbOTdhswubr9TL1cTjSAKWb7ZCNlsmDbTHrKObgEbEGTnaaqf6vrc1PdtN6qL62ZcHZrgy6w26LQMfpZY5R4-7AynQ+Goc4YR5vPDEkibWQAPS1qDCYDELKixBwZBkYo1EAHKCtzAdib5x1ajjkPkChMfDNWkVirqcAo1GqcRWY+uN5tQNsdrs9vu75DD9MB0fSCdUFXAE+J92WhgsecVCadACM66AA)

Reply:
Note that in my example the `ValueTypeForFieldValueType` interface is not necessarily true, I just put what I think it might be without checking.

Reply:
This is really cool I didn't know we could do this with TS 🤯